feat: server-side rendering of styled tiles

CommanderStorm opened the pull request:
### Updates by @nyurik

To test
* modify `render/cfg.yml` to point to some local openmaptiles-compatible .mbtiles file, either some area or the whole planet
* run `cargo run -- --webui enable-for-all --config render/cfg.yml`
* open `render/debug.html` in a browser (as local file, without any servers)

Current result: **segmentation fault**

What already works:  http://localhost:3000/style/bright/0/0/0.png and any other location should show an image, as long as there is only one simultaneous request

### Original comments by @CommanderStorm
This PR is a quite naive draft at how we could integrate maplibre_native.
It assumes that styling is done and mergable => builds on top of that base.
It also assumess that the challenge in this PR will be getting the build system to cooperate => includes a quite simple approach to just build up and tear down a renderer for each request.

- Fixes: #1685 
- Fixes: #978

TODOs (upstream):
- <details><summary>fix the git submodule issue (click to expand)</summary>
    <p>
    
    ![image](https://github.com/user-attachments/assets/59bbf3e5-020b-497a-a2fa-3906726aa6f4)
    
    </p>
    </details> 
- <details><summary>fix the git-clone issue on rebuilds (click to expand)</summary>
  <p>
  
  ![image](https://github.com/user-attachments/assets/fd5d4645-28dd-462a-96a5-dd12516ea6b6)
  
  
  </p>
  </details> 


TODOs (here):
- [ ] integrate with docker (=> make sure that all the things are installed)
- [ ] integrate with CI (=> make sure that all the things are installed)
- [ ] add a new testcase that does not have rendering feature-flagged, but should still build and test
- [ ] integrate with our testsuite (once it can be build)

There are a few things, where I think pushing to future PRs is a better call, after the intital challenges of the build system.
- [ ] Is `maplibre_native` threadsafe? => how should said renderer be pooled
- [ ] How many renderers should there be? How should it scale to more load
- [ ] Which type of renderers should there be (static, tiles, interactive)

CommanderStorm committed:
basic tile rendering

fix clippy issue

CommanderStorm committed:
feat: style serving

This PR adds style serving to the features we support.
Resolves #1607

I am not quite certain if `sources` is something that works for styles.. The key to the dict not playing any role is somewhat counterintuitive.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
merge fixes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm committed:
Merge branch 'main' into style-rendering

nyurik committed:
Merge branch 'main' into style-rendering

nyurik committed:
wip

nyurik committed:
fixes

nyurik committed:
add a tester script - do not merge!

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyurik committed:
Change init map zoom level to 0 in debug

this prevents a crash from the start

CommanderStorm committed:
renamed the feature to "render-styles"

CommanderStorm committed:
don't track the cache

CommanderStorm committed:
add a big fat mutex in front

nyurik commented on martin-core/Cargo.toml:
```diff
@@ -39,7 +39,7 @@ sprites = [
     "dep:dashmap",
 ]
-render = ["styles", "dep:maplibre_native"]
+render-styles = ["styles", "dep:maplibre_native"]
```
why styles? I feel `render-styles` is a bit confusing tbh

CommanderStorm replied:
why confusing?
I think it makes more sense than just `render`..

Both likely work, I am happy to revert that one if you have a strong oppinion.

nyurik replied:
are there any other type of "rendering" that we would do in a tile server?

CommanderStorm replied:
Really not sure as said above.
I don't think another usecase exitsts.

I am more thinking about what `cargo install martin --no-default-features -F rendering` vs `cargo install martin --no-default-features -F style-endering` feels like.

nyurik replied:
The common name used for this is `server-side rendering` - thus `style-rendering` is a bit confusing. Since Martin is a server anyway, might as well use `rendering` or `render` for simplicity.

CommanderStorm committed:
optimise sligtly by switching to a lazy, static, rendering actor-thread

CommanderStorm committed:
update deps to make merge with main more pleaseant

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
revert merge issue

CommanderStorm commented on docs/src/sources-styles.md:
```diff
@@ -21,5 +21,4 @@ A restart of Martin is required to see new styles.
 
-This is not implemented yet, but there is a plan to add it.
-Please see <https://github.com/maplibre/martin/issues/978> for more information.
+TODO, document
```
@nyurik what kind of configuration scheme would we like?
How about a simple boolean for now?

Should this default to `true|false`?

```yaml
# Publish MapLibre style files for client and server side rendering
styles:
   # EXPERIMENTAL - Enable static, server side, style rendering
   #
   # We are not currently happy with the performance of this endpont and plan on improving this in the future
   allow_static_rendering: true|false

   paths:
     # publish all *.json files in this directory
     # The name of the file will be used as the style name
     - /path/to/styles_dir
     # publish a single file - here `maplibre_style` will be the style name
     - /path/to/maplibre_style.json
   sources:
     # publish a JSON file found at this path as `some_style_name`
     #
     # Contrairy to paths, if directories are specified, Martin will print a
     # warning and ignore them.
     # To serve a style-directory, use the `paths` section above or name each
     # style individually. This prevents footguns with names being unclear.
     some_style_name: /path/to/this/style.json
     #  Publish specific file as `other_style_name`
     other_style_name: /path/to/other_style.json
```

nyurik replied:
* snapshots:   `zoom`, `center lat`, `center lon`, `img width`, `img height`, DPI `1x / 2x / 3x / ...` -- enabled/max img size/allowed DPIs
* tile - enabled, allowed-DPIs

we don't have to do all these at first, just thinking out lout. Also, take a look at the tileserve-gl -- see what configs they have

CommanderStorm replied:
I am not sure if I sync on what snapshots are.
Do you mean "generate a map of this area, overlay a marker/route/.."?

In this case, most of these should be runtime options => `zoom`, `center`, `width`/`height`, DPI `@1x..`, different image format (webp).
I intend there to be a second API for generating these static images.
The current API is optimised for a different usecase.
Tileserver-gl has a [fairly nice API](https://tileserver.readthedocs.io/en/latest/endpoints.html#static-images) as it seems.

CommanderStorm replied:
In terms of usage limits, we likely need a few of those (`max_zoom`, …)
Not sure which we exactly require though to be able to get this feature out the door.

Since we are not going to have very option after this PR, lets change tactics and move more iteratively.
=> the schema just has to be extendable

How about:

```yaml
# Publish MapLibre style files for client and server side rendering
styles:
   # EXPERIMENTAL - static, server side, style rendering
   #
   # We are not currently happy with the performance of this endpoint and intend to improve this in the future
   # Marking this experimental means that we are not stuck with single threaded performance as a default until v2.0
   rendering: # can also be "rendering: false|true" for easy defaults
        # Impl in another PR. Lets keep things review + testable
       max_zoom: 20
     
```

nyurik replied:
i agree with the structure - `rendering: true/false/object` where the object is TBD.   Do note that for now we may want to rename it to `experimental_rendering: false` (disabled by default)

CommanderStorm replied:
Alternatively, we could do something like:

```yaml
# Publish MapLibre style files for client and server side rendering
styles:
   # EXPERIMENTAL - static, server side, style rendering
   #
   # We are not currently happy with the performance of this endpoint and intend to improve this in the future
   # Marking this experimental means that we are not stuck with single threaded performance as a default until v2.0
   rendering:
       tiled: true 
       static: true 
       limits: {}

nyurik replied:
meh, just call it `render-experimental: true` (false by default).  Baby steps :)

CommanderStorm replied:
If we do it this way we require changes once we stabilise.
I would prefer to not break like this and instead just mark it in the docs as experimental.

nyurik replied:
i am ok to break it when stabilized. Either that, or we can introduce `--enable-experimental-render` CLI param. Noone looks at the docs, but printing a warning that "experimental feature ... is enabled" and requiring it to be clearly marked as experimental is a rust-like approach that would prevent any wrong expectations.

CommanderStorm replied:
`--enable-experimental-render` seems to need a larger refactoring. I went with the config option instead.

CommanderStorm committed:
use a oneshot channel for slightly more optimal performance

CommanderStorm committed:
fix formatting

CommanderStorm committed:
optimise the `set_style_path` usage

CommanderStorm committed:
fix clippy lint

CommanderStorm committed:
Write integration and unit tests for rendering

CommanderStorm committed:
add an e2e test

CommanderStorm committed:
update `maplibre_native`

CommanderStorm committed:
fix clippy issues

CommanderStorm committed:
add cfg checks

CommanderStorm committed:
fix a testcase cfg being wrong

CommanderStorm committed:
preemtively switch to using `rendering as a fewature name`

CommanderStorm committed:
unify config handling

CommanderStorm committed:
Create error.rs

CommanderStorm committed:
mend

CommanderStorm committed:
Update save_config.yaml

CommanderStorm committed:
fix clippy lint

CommanderStorm committed:
document the feature better

CommanderStorm committed:
make sure that just is installed in CI

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
fix just not being installed for codecov

CommanderStorm committed:
migrate to improved api

CommanderStorm committed:
migrate to using the pool

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyurik commented on martin/src/config/file/resources/styles.rs:
```diff
@@ -12,11 +14,50 @@ use crate::config::file::{
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct RendererConfig {
+    // Same effect as experimental_rendering: true|false shorthands
+    enabled: bool,
```
I think we may need one more thing here - a path to sqlite file - creating sqlite in a current dir might cause all sorts of issues in production. Eventually i do hope we can get rid of the whole sqlite thing, but for now ... sigh

CommanderStorm replied:
Not as simple to add under the architecture I manoeuvred us into.
Lets move this to another PR

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
add the install commands to CI

CommanderStorm committed:
change to the proper MLN-rs version

CommanderStorm committed:
more ci wrangling

CommanderStorm commented on .github/workflows/bench.yml:
```diff
@@ -23,5 +23,16 @@ jobs:
+            libcurl4-openssl-dev \
+            libglfw3-dev \
+            libuv1-dev \
+            libz-dev
```
```suggestion
      - run: rustup update stable && rustup default stable
      - uses: taiki-e/install-action@cc60de1d6831d7e9c4342f618ce7a5d6a9f223a4 # v2.61.6
        with: { tool: 'just' }
      - name: Install dependencies for rendering
        run: just install-dependencies
```

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
refactor location of rendering path

CommanderStorm committed:
Create styles_rendering.rs

CommanderStorm committed:
remove debugging files

CommanderStorm committed:
revert lockfile changes

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
fix(docs): publish docs to docs.rs again (#2239)

I have the hunch that this might work.
Not quite sure, but at least from https://docs.rs/about/builds this is
how this is done

CommanderStorm committed:
chore(release): improve release-plz config (#2242)

This PR fixes a few bugs in our release plz config:
- martins changelog not containing its own changes
- dependency versions being bumped wrong

CommanderStorm committed:
chore(release): bump pmtiles (#2232)

this PR bumps our pmtiles version to the next major one

CommanderStorm commented on .github/workflows/ci.yml:
```diff
@@ -114,16 +124,19 @@ jobs:
       - uses: taiki-e/install-action@cc60de1d6831d7e9c4342f618ce7a5d6a9f223a4 # v2.61.6
-        with: { tool: cargo-zigbuild }
-      - run: rustup target add ${{ matrix.target }}
+        with: { tool: cargo-zigbuild,just }
```
```suggestion
        with: { tool: 'cargo-zigbuild,just' }
```

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
revert a few unrelated changes

nyurik commented on .github/workflows/ci.yml:
```diff
@@ -360,19 +373,24 @@ jobs:
         uses: ilammy/setup-nasm@72793074d3c8cdda771dba85f6deafe00623038b # v1.5.2
+      - name: Install dependencies for rendering (Linux)
+        if: runner.os == 'Linux'
+        run: just install-dependencies
```
doesn't the `install-dependencies` recipe automatically use the right deps for the OS? Why filter for linux here?

CommanderStorm replied:
We don't have the recipe for windows, also for macos this needs another pass.

nyurik replied:
We can create an empty one for windows. What do you mean another pass - we should look at it again, or we need to run it twice?

CommanderStorm replied:
maplibre-native-rs does not currently support macos.
The integration there needs another pass.

I don't care either way, adding the empty recipy now

CommanderStorm committed:
change install-dependencies to work under windows too

nyurik commented on .github/workflows/ci.yml:
```diff
@@ -441,6 +460,14 @@ jobs:
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+      - name: Set up gdal-bin and sqlite3-tools
+        if: runner.os == 'Linux'
+        run: sudo apt-get update && sudo apt-get install -y gdal-bin sqlite3-tools
```
should gdal be also part of install dependencies recipe? just curious, ok if not

CommanderStorm replied:
Since our long term plan for this is to add a new crate to do the e2e tests in, lets keep it as is.

nyurik committed:
ignore cache.sqlite-journal

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
mark the rendering feature as unstable and default-disable it

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm committed:
Merge branch 'main' into style-rendering

CommanderStorm commented on tests/test.sh:
```diff
@@ -766,3 +773,20 @@ fi
+
+rm -rf "$TEST_TEMP_DIR"
+
+>&2 echo "All integration tests have passed"
```
```suggestion
```

CommanderStorm commented on martin/src/srv/server.rs:
```diff
@@ -117,6 +117,9 @@ pub fn router(cfg: &mut web::ServiceConfig, #[allow(unused_variables)] usr_cfg:
 
+    #[cfg(all(feature = "unstable-rendering", target_os = "linux"))]
+    cfg.service(crate::srv::styles_rendering::get_style_rendered);
+
```
```suggestion
```

CommanderStorm commented on martin/Cargo.toml:
```diff
@@ -113,6 +113,7 @@ postgres = [
     "dep:postgres-protocol",
     "_tiles",
 ]
+unstable-rendering = ["styles", "martin-core/unstable-rendering", "dep:image"]
```
```suggestion
```

CommanderStorm commented on martin/martin-ui/package-lock.json:
```diff
@@ -1,3 +1,4 @@
+
```
```suggestion
```

CommanderStorm commented on martin-core/Cargo.toml:
```diff
@@ -50,6 +50,7 @@ sprites = [
     "dep:futures",
     "dep:dashmap",
 ]
+unstable-rendering = ["styles", "dep:maplibre_native", "dep:image"]
```
```suggestion
```

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm commented on martin/src/config/file/tiles/postgres/config.rs:
```diff
@@ -252,8 +252,7 @@ impl ConfigurationLivecycleHooks for PostgresConfig {
                     .iter()
-                    .map(|k| format!("auto_publish.{k}"))
-                    .collect::<UnrecognizedKeys>(),
+                    .map(|k| format!("auto_publish.{k}")),
```
```suggestion
                    .map(|k| format!("auto_publish.{k}"))
                    .collect::<UnrecognizedKeys>(),
```

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm committed:
remove files from the diff
